### PR TITLE
[charts] Fix pie chart disappearing when an arc spans almost the full circle

### LINKED ISSUE #14167: [charts] PieChart disappears with certain data values
### Steps to reproduce

Link to live example: https://codesandbox.io/p/sandbox/epic-bush-4jk7jf

This issue may have been raised before, however, the example shows a pie chart hidden and 2 pie charts that are visible.

Screenshot from 2024-08-12 10-10-51

### Current behavior

The pie chart is hidden

### Expected behavior

The pie chart is visible

### Context

My use case is the data values are dynamic and have ran into this issue by accident

### Your environment

_No response_

**Search keywords**: piechart

### PR DESCRIPTION
**Pull request by JCQuintas**
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Fixes #14167

When a pie slice is extremely small compared to the largest one (e.g. `[999_999, 1]`), the large arc spans almost, but not exactly, the full circle. Its start and end points then round to the same coordinates in the serialized path (d3-shape uses a precision of 3 digits), and per the SVG spec an arc whose end point equals its start point is omitted — the quasi-full-circle slice vanishes, making the whole pie look invisible. The failing zone is radius-dependent, which is why resizing the chart by 1px could toggle the bug.

This PR snaps the end angle of such arcs to a true full circle when the gap they leave is smaller than the path precision (0.002px chord), so d3 emits its two-arc full-circle form. The snapping accounts for radius, direction, and pad angle (padded arcs are never wrongly snapped), and only applies when the gap would be invisible anyway. Applied in `useAnimatePieArc`, covering the plot, the focused arc, and all animation frames. A visual regression fixture with formerly-hidden and visible value ratios is included.

## Changelog

Fixed pie charts rendering as empty when one slice spans almost the full circle.

### CODE REVIEW

File: packages/x-charts/src/internals/snapEndAngleToFullCircle.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+const TAU = 2 * Math.PI;
+
+/* `d3-shape` serializes paths with a precision of 3 digits, i.e., a resolution of 1e-3. */
+const PATH_PRECISION = 1e-3;
+/* If the distance between the start and end points of an arc is at least twice the path
+ * precision, at least one of their coordinates is guaranteed to round to a different value. */
+const MIN_DISTINCT_CHORD_LENGTH = 2 * PATH_PRECISION;
```
Comment (alexfauquette) [inline packages/x-charts/src/internals/snapEndAngleToFullCircle.ts:7]: I wondered where this `2` came from :)

Technically you could go up to sqrt(2). But 2 is good enought :+1:

Image

File: test/regressions/charts/PieNearFullCircle.tsx
Comment (alexfauquette) [inline test/regressions/charts/PieNearFullCircle.tsx:1]: Not sure it's worth having a screen shot for each PR if we can test this edge case with JSdom

**Review by alexfauquette — APPROVED** [review]
Thanks for the investigation. I did not know the D3 rounding was set to 0.001

File: test/regressions/charts/PieNearFullCircle.tsx
Comment (JCQuintas) [reply test/regressions/charts/PieNearFullCircle.tsx:1]: easier to understand/notice a regression with the image test. Specially with AI, sometimes it might change the test to make it pass because of a new feature but reintroduce the bug again 🫠

Comment (alexfauquette) [reply test/regressions/charts/PieNearFullCircle.tsx:1]: >  it might change the test to make it pass

😭
